default_validator/src/lib.rs

```diff
@@ -33,107 +31,176 @@ macro_rules! map_all_crate_tokens {
     };
 }
 
-pub(crate) use map_all_crate_tokens;
+macro_rules! tokens {
+    (
+        pattern = {
+            $(#[$meta:meta])*
+            $vis:vis struct _ {
+                $(
+                    $(#[$field_meta:meta])*
+                    $field_vis:vis $field:ident: $field_type:ty
+                ),* $(,)?
+            }
+        },
+        $module:ident :: tokens: [$($name:ident),+ $(,)?]
+    ) => {
+        declare_tokens!($(
+            crate::isi::$module::tokens::$name
+        ),+);
+
+        pub mod tokens {
+            //! Permission tokens for concrete operations.
+
+            use super::*;
+
+            macro_rules! single_token {
```

Comment:
Why do we need macros here?

Reply:
@Arjentix ?

Reply:
It was the only thing I came out here to make `tokens!` macro working like I wanted. Although I'm agree that its code and the usage code aren't very readable. The main pros of `tokens!` macro is that it significantly reduces the code size by eleminating repetative things.

So basically what `tokens!` does is declaring multiple `PermissionTokenDefinition`s. It's usable when all tokens are diffirent only in names, that's why we have *pattern* token input. `single_token!` inside `tokens!` declares a single token which will look like described in *pattern*. And then we call it for all token names provided to `tokens!` macro

Reply:
It's a workaround for problem then you can't iterate other things with different length inside declarative macro